File: 2015/day16/day16.py
#!/usr/bin/env python3

import re
import sys

filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"

aunts = {}

for line in open(filename):
    m = re.match(r"Sue (\d+): (\w+): (\d+), (\w+): (\d+), (\w+): (\d+)", line)
    aunts[m.group(1)] = {m.group(2): int(m.group(3)), m.group(4): int(m.group(5)), m.group(6): int(m.group(7))}

for sue, aunt in aunts.items():
    if aunt.get("children", 3) != 3:
        continue
    if aunt.get("cats", 7) != 7:  # should be greater than
        continue
    if aunt.get("samoyeds", 2) != 2:
        continue
    if aunt.get("pomeranians", 3) != 3:  # should be fewer than
        continue
    if aunt.get("akitas", 0) != 0:
        continue
    if aunt.get("vizslas", 0) != 0:
        continue
    if aunt.get("goldfish", 5) != 5:  # should be fewer than
        continue
    if aunt.get("trees", 3) != 3:  # should be greater than
        continue
    if aunt.get("cars", 2) != 2:
        continue
    if aunt.get("perfumes", 1) != 1:
        continue

    print(sue)


for sue, aunt in aunts.items():
    if aunt.get("children", 3) != 3:
        continue
    if aunt.get("cats", 7 + 1) <= 7:  # should be greater than
        continue
    if aunt.get("samoyeds", 2) != 2:
        continue
    if aunt.get("pomeranians", 3 - 1) >= 3:  # should be fewer than
        continue
    if aunt.get("akitas", 0) != 0:
        continue
    if aunt.get("vizslas", 0) != 0:
        continue
    if aunt.get("goldfish", 5 - 1) >= 5:  # should be fewer than
        continue
    if aunt.get("trees", 3 + 1) <= 3:  # should be greater than
        continue
    if aunt.get("cars", 2) != 2:
        continue
    if aunt.get("perfumes", 1) != 1:
        continue

    print(sue)

File: 2015/day16/day16.rs
//! [Day 16: Aunt Sue](https://adventofcode.com/2015/day/16)

use regex::Regex;
use rustc_hash::FxHashMap;

struct Puzzle<'a> {
    aunts: FxHashMap<u32, FxHashMap<&'a str, u32>>,
}

impl<'a> Puzzle<'a> {
    fn new(data: &'a str) -> Self {
        let mut aunts: FxHashMap<u32, FxHashMap<&'a str, u32>> = FxHashMap::default();

        let re = Regex::new(r"Sue (\d+): (\w+): (\d+), (\w+): (\d+), (\w+): (\d+)").unwrap();

        for line in data.lines() {
            let m = re.captures(line).unwrap();
            let sue = m.get(1).unwrap().as_str().parse::<u32>().unwrap();

            for i in (2..=6).step_by(2) {
                let key = m.get(i).unwrap().as_str();
                let value = m.get(i + 1).unwrap().as_str().parse::<u32>().unwrap();
                aunts.entry(sue).or_default().insert(key, value);
            }
        }
        Self { aunts }
    }

    fn part1(&self) -> u32 {
        for (sue, aunt) in &self.aunts {
            if aunt.get("children").unwrap_or(&3) == &3
                && aunt.get("cats").unwrap_or(&7) == &7
                && aunt.get("samoyeds").unwrap_or(&2) == &2
                && aunt.get("pomeranians").unwrap_or(&3) == &3
                && aunt.get("akitas").unwrap_or(&0) == &0
                && aunt.get("vizslas").unwrap_or(&0) == &0
                && aunt.get("goldfish").unwrap_or(&5) == &5
                && aunt.get("trees").unwrap_or(&3) == &3
                && aunt.get("cars").unwrap_or(&2) == &2
                && aunt.get("perfumes").unwrap_or(&1) == &1
            {
                return *sue;
            }
        }
        0
    }

    fn part2(&self) -> u32 {
        for (sue, aunt) in &self.aunts {
            if aunt.get("children").unwrap_or(&3) == &3
                && aunt.get("cats").unwrap_or(&8) > &7          // should be greater than
                && aunt.get("samoyeds").unwrap_or(&2) == &2
                && aunt.get("pomeranians").unwrap_or(&2) < &3   // should be fewer than
                && aunt.get("akitas").unwrap_or(&0) == &0
                && aunt.get("vizslas").unwrap_or(&0) == &0
                && aunt.get("goldfish").unwrap_or(&4) < &5      // should be fewer than
                && aunt.get("trees").unwrap_or(&4) > &3         // should be greater than
                && aunt.get("cars").unwrap_or(&2) == &2
                && aunt.get("perfumes").unwrap_or(&1) == &1
            {
                return *sue;
            }
        }

        0
    }
}

/// main function
fn solve(data: &str) -> (u32, u32) {
    let puzzle = Puzzle::new(data);
    (puzzle.part1(), puzzle.part2())
}

fn main() {
    let mut args = aoc::parse_args();
    args.run(solve);
}
